File: src/main/java/org/ant/plugin/ConnectFour.java
package org.ant.plugin;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.*;
import org.bukkit.block.Block;
import org.bukkit.configuration.serialization.ConfigurationSerializable;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitTask;

import java.util.HashMap;
import java.util.Map;

public class ConnectFour implements ConfigurationSerializable {
    Location location;
    Location center;
    String align;

    int[][] board;
    int[] top;
    int selected = -1;
    BukkitTask display_selected_task;
    int player;
    boolean end;
    Player[] minecraft_players;

    public ConnectFour(Location location, String align) {
        this.location = location;
        this.align = align;
        if(align.equals("x"))this.center = location.clone().add(3.5, 3, 0);
        else if(align.equals("z"))this.center = location.clone().add(0, 3, 3.5);
        reset();
    }
    public void reset(){
        board = new int[6][7];
        top = new int[7];
        selected = -1;
        if(display_selected_task != null)display_selected_task.cancel();
        player = 1;
        minecraft_players = new Player[2];
        end = false;
        for(int x=0; x<7; x++){
            for(int y=0; y<7; y++){
                int status = 0;
                if(x == 0)status = -1;
                if(align.equals("x"))location.clone().add(y,x,0).getBlock().setType(Method.yellow_red_material(status));
                else if(align.equals("z"))location.clone().add(0,x,y).getBlock().setType(Method.yellow_red_material(status));
            }
        }
    }

    

    boolean visible = true;
    public boolean move(int y, Player minecraft_player) {
        if(!end && top[y] < 6){
            if(minecraft_players[player - 1] == null || minecraft_players[player - 1].equals(minecraft_player)) {
                if (minecraft_players[player - 1] == null) minecraft_players[player - 1] = minecraft_player;
                if (selected == -1 || selected != y) {
                    if (display_selected_task != null) display_selected_task.cancel();
                    if (selected != -1) {
                        Block block = null;
                        if (align.equals("x")) block = location.clone().add(selected, 0, 0).getBlock();
                        else if (align.equals("z")) block = location.clone().add(0, 0, selected).getBlock();
                        block.setType(Method.yellow_red_material(-1));
                    }
                    selected = y;
                    visible = true;
                    display_selected_task = Bukkit.getScheduler().runTaskTimer(Game.getInstance(), () -> {
                        Block selected_block = null;
                        if (align.equals("x")) selected_block = location.clone().add(y, 0, 0).getBlock();
                        else if (align.equals("z")) selected_block = location.clone().add(0, 0, y).getBlock();

                        if (visible) selected_block.setType(Method.yellow_red_material(player));
                        else selected_block.setType(Method.yellow_red_material(-1));
                        visible = !visible;
                    }, 0, 10);
                } else {
                    display_selected_task.cancel();
                    Block block = null;
                    if (align.equals("x")) block = location.clone().add(y, 0, 0).getBlock();
                    else if (align.equals("z")) block = location.clone().add(0, 0, y).getBlock();
                    block.setType(Method.yellow_red_material(-1));
                    selected = -1;

                    board[top[y]][y] = player;
                    if (align.equals("x"))
                        location.clone().add(y, 6, 0).getBlock().setType(Method.yellow_red_material(player));
                    else if (align.equals("z"))
                        location.clone().add(0, 6, y).getBlock().setType(Method.yellow_red_material(player));
                    if (is_win(top[y], y)) {
                        Component component;
                        if (player == 1) {
                            component = Component.text("黃色勝利").color(NamedTextColor.YELLOW);
                            Method.yellow_red_firework(center.clone().add(0, 3, 0), true);
                        } else {
                            component = Component.text("紅色勝利").color(NamedTextColor.RED);
                            Method.yellow_red_firework(center.clone().add(0, 3, 0), false);
                        }
                        Method.broadcast(component, center, 7);
                        end = true;
                    } else if (is_tie()) {
                        Method.broadcast("平手", center, 7);
                        end = true;
                    } else {
                        top[y]++;
                        if (player == 1) player = 2;
                        else player = 1;
                    }
                }
                return true;
            }
            else{
                Component component;
                component = Component.text("已被玩家 " + minecraft_players[player - 1].getName() + " 綁定").color(NamedTextColor.RED);
                minecraft_player.sendMessage(component);
                minecraft_player.playSound(minecraft_player, Sound.ENTITY_VILLAGER_NO, 1.0f, 1.0f);
                return false;
            }
        }
        return false;
    }

    public boolean is_inside(int x, int y){
        return x >= 0 && x < 6 && y >= 0 && y < 7;
    }
    private boolean is_win(int x, int y){
        int[][] vectors = {{1,0},{0,1},{1,1},{1,-1},{-1,0},{0,-1},{-1,-1},{-1,1}};
        int[] counter = new int[4];
        for(int i = 0; i <8; i++){
            int check_x = x, check_y = y;
            for(int j = 1; j <=3; j++){
                check_x += vectors[i][0];
                check_y += vectors[i][1];
                if(is_inside(check_x, check_y)){
                    if(board[check_x][check_y] == player){
                        counter[i%4]++;
                    }
                    else break;
                }
                else break;
            }
            if(counter[i%4] >=3)return true;
        }
        return false;
    }
    private boolean is_tie(){
        for(int x=0; x<6; x++){
            for(int y=0; y<7; y++){
                if(board[x][y] == 0)return false;
            }
        }
        return true;
    }

    public Map<String, Object> serialize() {
        Map<String, Object> data = new HashMap<>();
        data.put("location", this.location);
        data.put("align", this.align);
        return data;
    }

    public static ConnectFour deserialize(Map<String, Object> args) {
        return new ConnectFour(
            (Location) args.get("location"),
            (String) args.get("align")
        );
    }
}

File: src/main/kotlin/org/ant/game/gameimpl/ConnectFour.kt
package org.ant.game.gameimpl

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.ant.game.AntGamePlugin
import org.ant.game.gameimpl.gameframe.BoardGame
import org.ant.game.gameimpl.gameframe.GameConstants
import org.ant.game.gameimpl.gameframe.GameDeSerializable
import org.ant.game.gameimpl.gameframe.GameSerializable
import org.ant.game.gameimpl.gameframe.GameState
import org.ant.game.gameimpl.gameframe.Method
import org.ant.game.gameimpl.gameframe.RecordSerializable
import org.ant.game.gameimpl.gameframe.UUIDPair
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Sound
import org.bukkit.block.Block
import org.bukkit.entity.Entity
import org.bukkit.entity.FallingBlock
import org.bukkit.entity.Player
import org.bukkit.event.entity.CreatureSpawnEvent
import org.bukkit.scheduler.BukkitTask
import org.bukkit.util.Vector

class ConnectFour(val pluginInstance: AntGamePlugin) :
    BoardGame(SIZE),
    GameSerializable,
    RecordSerializable {
    companion object : GameDeSerializable {
        const val SIZE = 7

        override fun deserialize(pluginInstance: AntGamePlugin, args: Map<String, Any?>): ConnectFour {
            val connectFour = ConnectFour(pluginInstance)

            @Suppress("UNCHECKED_CAST")
            val boards = args["boards"] as Map<String, Map<String, Any?>>
            for ((key, value) in boards) {
                connectFour.setBoard(
                    key,
                    value["origin"] as Location,
                    value["xAxis"] as Vector,
                    value["yAxis"] as Vector
                )
            }
            return connectFour
        }
    }
    var boardState = Array(SIZE) { IntArray(SIZE) }
    var selected: Int = -1
    var displaySelectedTask: BukkitTask? = null
    val fallingBlocks = arrayListOf<Entity>()
    var player: Int = 1
    var end: Boolean = false
    val uuidPair = UUIDPair()

    var visible: Boolean = true

    init {
        reset(null)
    }

    override fun serialize(): MutableMap<String, Any?> {
        val data = hashMapOf<String, Any?>()
        for ((name, board) in boards) {
            data["boards.$name"] = mapOf<String, Any?>(
                "origin" to board.origin,
                "xAxis" to board.xAxis,
                "yAxis" to board.yAxis
            )
        }
        return data
    }

    override fun deserializeRecord(data: Map<String, Any?>) {
        @Suppress("UNCHECKED_CAST")
        reset(
            GameState(
                Method.deserialize2dBoard(data["board"] as List<String>?),
                data["player"] as Int,
                data["end"] as Boolean
            )
        )
    }

    override fun serializeRecord(): MutableMap<String, Any?> {
        val data = HashMap<String, Any?>()
        data["board"] = Method.serialize2dBoard(boardState)
        data["player"] = player
        data["end"] = end

        return data
    }

    override fun reset(gamePreset: GameState?) {
        if (gamePreset != null) {
            @Suppress("UNCHECKED_CAST")
            boardState = gamePreset.boardState as Array<IntArray>
            player = gamePreset.player
            end = gamePreset.end
        } else {
            boardState = Array(SIZE) { IntArray(SIZE) }
            selected = -1
            player = 1
            end = false
        }

        displaySelectedTask?.cancel()
        uuidPair.clear()
        display()
    }

    override fun move(x: Int, y: Int, z: Int, movePlayer: Player): Boolean {
        for (fallingBlock in fallingBlocks) {
            if (fallingBlock.isValid) return false
        }
        fallingBlocks.clear()
        if (!end && boardState[x][6] < 6) {
            val playerUUID = movePlayer.uniqueId
            if (selected != -1) {
                for (board in boards.values) {
                    board.origin.clone().add(board.xAxis.clone().multiply(selected)).block.type = Method.yellowRedMaterial(-1)
                }
            }
            if (uuidPair.getPlayerUUID(player) == playerUUID || uuidPair.putPlayerUUID(playerUUID)) {
                displaySelectedTask?.cancel()
                if (selected == -1 || selected != x) {
                    selected = x
                    visible = true
                    displaySelectedTask = Bukkit.getScheduler().runTaskTimer(
                        pluginInstance,
                        Runnable {
                            for (board in boards.values) {
                                val selectedBlock: Block = board.origin.clone().add(board.xAxis.clone().multiply(x)).block
                                if (visible) {
                                    selectedBlock.type = Method.yellowRedMaterial(player)
                                } else {
                                    selectedBlock.type = Method.yellowRedMaterial(-1)
                                }
                                visible = !visible
                            }
                        },
                        0,
                        10
                    )
                } else {
                    display()
                    selected = -1
                    boardState[x][boardState[x][6]] = player
                    for (board in boards.values) {
                        val location = board.origin.clone()
                            .add(board.xAxis.clone().multiply(x))
                            .add(board.yAxis.clone().multiply(6))
                            .add(0.5, 0.0, 0.5)
                        fallingBlocks.add(
                            location.world.spawn(location, FallingBlock::class.java, CreatureSpawnEvent.SpawnReason.DEFAULT) { entity ->
                                entity.blockData = Method.yellowRedMaterial(player).createBlockData()
                                entity.dropItem = false
                                entity.velocity = Vector(0.0, 0.1, 0.0)
                            }
                        )
                    }
                    if (isWin(x, boardState[x][6])) {
                        val component: Component
                        if (player == 1) {
                            component = Component.text("黃色勝利").color(NamedTextColor.YELLOW)
                            for (board in boards.values) {
                                Method.yellowRedFirework(board.center.clone().add(0.0, 4.0, 0.0), true)
                            }
                        } else {
                            component = Component.text("紅色勝利").color(NamedTextColor.RED)
                            for (board in boards.values) {
                                Method.yellowRedFirework(board.center.clone().add(0.0, 4.0, 0.0), false)
                            }
                        }
                        broadcast(component)
                        end = true
                    } else if (isTie()) {
                        broadcast("平手")
                        end = true
                    } else {
                        boardState[x][6]++
                        player = if (player == 1) {
                            2
                        } else {
                            1
                        }
                    }
                }
                return true
            } else {
                val component = Component.text("已被玩家 " + Bukkit.getPlayer(uuidPair.getPlayerUUID(player)!!)?.name + " 綁定").color(NamedTextColor.RED)
                movePlayer.sendMessage(component)
                movePlayer.playSound(movePlayer, Sound.ENTITY_VILLAGER_NO, 1.0f, 1.0f)
                return false
            }
        }
        return false
    }

    override fun display() {
        for (board in boards.values) {
            for (x in 0..<SIZE) {
                val location = board.origin.clone().add(board.xAxis.clone().multiply(x))

                location.block.type = Method.yellowRedMaterial(-1)
                location.add(board.yAxis)
                for (y in 0..<6) {
                    location.block.type = Method.yellowRedMaterial(boardState[x][y])
                    location.add(board.yAxis)
                }
            }
        }
    }

    private fun isWin(x: Int, y: Int): Boolean {
        val counter = IntArray(4)
        for (i in 0..7) {
            var checkX = x
            var checkY = y
            @Suppress("UNUSED_PARAMETER")
            for (j in 1..3) {
                checkX += GameConstants.eightDirection[i][0]
                checkY += GameConstants.eightDirection[i][1]
                if (isInside(checkX, checkY) && y < 6) {
                    if (boardState[checkX][checkY] == player) {
                        counter[i % 4]++
                    } else {
                        break
                    }
                } else {
                    break
                }
            }
            if (counter[i % 4] >= 3) return true
        }
        return false
    }

    private fun isTie(): Boolean {
        for (x in 0..<SIZE) {
            for (y in 0..<6) {
                if (boardState[x][y] == 0) {
                    return false
                }
            }
        }
        return true
    }
}
